sph/include/sph/hydro_std/iad_kern.hpp

```diff
@@ -42,27 +45,31 @@ HOST_DEVICE_FUN inline void IADJLoopSTD(cstone::LocalIndex i, T sincIndex, T K,
 
         T mj_roj_w = m[j] / rho[j] * w;
 
-        tau11 += rx * rx * mj_roj_w;
-        tau12 += rx * ry * mj_roj_w;
-        tau13 += rx * rz * mj_roj_w;
-        tau22 += ry * ry * mj_roj_w;
-        tau23 += ry * rz * mj_roj_w;
-        tau33 += rz * rz * mj_roj_w;
+        tau[I::_11] += rx * rx * mj_roj_w;
+        tau[I::_12] += rx * ry * mj_roj_w;
+        tau[I::_13] += rx * rz * mj_roj_w;
+        tau[I::_22] += ry * ry * mj_roj_w;
+        tau[I::_23] += ry * rz * mj_roj_w;
+        tau[I::_33] += rz * rz * mj_roj_w;
     }
 
-    T det = tau11 * tau22 * tau33 + T(2) * tau12 * tau23 * tau13 - tau11 * tau23 * tau23 - tau22 * tau13 * tau13 -
-            tau33 * tau12 * tau12;
+    T normalization = T(1) / min(abs(tau));
```

Comment:
Why not use as normalization a value calculated at the beginning of the simulation, equal to the largest particle mass times pi times the largest coordinate value?

max_coord = max(max(x), max(y), max(z))
normalization = 1 / (max(m) * pi * max_coord)

I usually add the pi factor because systems may grow beyond their original size (i.e. explosions) and I don't want to make an exact multiple of the size.

Reply:
Yeah that was indeed a silly suggestion on my part. `normalization = 1 / (max(m) * pi * max_coord)` would work well, but we don't have max(m) at the moment. Computing it on the fly would be too expensive and computing it at the start of the simulation requires additional arguments to be passed down to the kernel.

Instead I'm proposing normalizing by the average exponent of the elements of tau. That brings back both small and large `tau`s closer to the origin, or leaves them more or less unchanged should some values be very big and some very small. Also avoids any division by zero problems.

Reply:
Your idea of averaging over exponents is good. That would work. But the truth to be told, we probably would never change the mass of the particles and if we do it won't be more than a few times (i.e. x2, x3). What if we simply take the mass of one particle at the beginning and we avoid all calculations?

Reply:
Once we have stars and dark matter, the masses might vary over a wider range compared to the cases that we're currently running. Looking at the average exponents is probably the most robust and also most adaptive.
In terms of cost, it's close to zero. Binary logarithms and raising a power of 2 are cheap.
Perhaps I could revert to the original `tau11`, `tau12` notation if you prefer. That would mean a very small code change. It seems the `tau` as array doesn't help so much anymore with the exponents.

Reply:
done